test/format/replay.c

```diff
@@ -0,0 +1,467 @@
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "format.h"
+
+/*
+ * Predictable replay is ability to do test runs multiple times and always have predictable changes
+ * made at every timestamp. Two predictable runs with the same starting data seed executed up to the
+ * same timestamp will always have their data compare identically. Predictable replay only works
+ * with timestamped transactions and to avoid complexity, only a single operation is allowed in a
+ * transaction.
+ *
+ * To achieve the predictability we use two random number generators (the data RNGs and the extra
+ * RNG) with known start seeds, the data seed and the extra seed. Every single-threaded modification
+ * (like bulk loading) when deciding on a random course, uses the global data RNG, which is seeded
+ * by the data seed. Global decisions that don't affect data, like whether to turn on verbose, or
+ * even the rate of checkpointing, use the global extra RNG, which is seeded by the extra seed.
+ * Changing the extra seed may change some characteristics of how a workload is tested, but should
+ * not change any data on disk. When worker threads run, they have their own data and extra RNGs,
+ * and these are seeded by the timestamp they are working on.
+ *
+ * Before a worker thread can decide on what operation to do on which key in which table, it must
+ * obtain the next timestamp. Timestamps are doled out atomically, so no two worker thread can ever
+ * perform operations using the same timestamp. The timestamp is XOR-ed with the data seed, the
+ * result is the seed of the thread's private data RNG for the duration of that operation. Likewise,
+ * a private extra RNG is seeded from the timestamp and the extra seed. This ensures that all
+ * decisions about what is committed at that timestamp is predictable based on the timestamp. As you
+ * might expect, the thread's data RNG is used to decide what operation to do, which table to use,
+ * and which key within the table. Other random decisions, like whether to reopen a session, or
+ * whether to repeat a read from the snap list, use the extra RNG.
+ *
+ * Note that once a thread has started to work on an operation at a timestamp, it cannot give up on
+ * the effort. If, for example, a rollback error naturally happens, we can rollback the transaction.
+ * However, immediately getting a new timestamp would mean that we would lose the consequences of
+ * the previous timestamp, perhaps a record would not be updated in a particular way. Thus, after a
+ * rollback, a thread starts again, using the same timestamp it had before, and it seeds its RNGs
+ * again using this timestamp. This gives full predictability, even in the face of temporary
```

Comment:
If a thread needs to do predictable operations within a transaction (and do the same operations another time, after a ROLLBACK), it will need to seed its RNG with the same value every time.  That value is based on the timestamp the thread is working on combined with the global seed.  At the time of the ROLLBACK, since it may have already used the RNG, it needs to reset it with the original seed it used.  Fortunately calculating the seed and resetting the RNG is a tiny operation.

Parts of this are covered in the previous paragraph.  Is there a better way to present it?

Reply:
I think the disconnect for me was that I assumed that selecting an operation meant getting a couple random numbers, using those to define the op, then executing the op. If I'm connecting the dots correctly, I think you're saying that as format performs the operation it may consult the RNG more times. 

I think the wording here is OK. Just clarifying my picture of what is happening.